tls/ttls/net.c

```diff
@@ -31,547 +32,227 @@
 
 /*
- * Read at most 'len' characters
+ * Accept a connection from a remote client
  */
-int mbedtls_net_recv( void *ctx, unsigned char *buf, size_t len )
+int
+mbedtls_net_accept(mbedtls_net_context *bind_ctx,
+		   mbedtls_net_context *client_ctx,
+		   void *client_ip, size_t buf_size,
+		   size_t *ip_len)
 {
-    int ret;
-    int fd = ((mbedtls_net_context *) ctx)->fd;
-
-    if( fd < 0 )
-        return( MBEDTLS_ERR_NET_INVALID_CONTEXT );
-
-    ret = (int) read( fd, buf, len );
-
-    if( ret < 0 )
-    {
-        if( net_would_block( ctx ) != 0 )
-            return( MBEDTLS_ERR_SSL_WANT_READ );
+	int ret;
+	struct socket *srv_socket = bind_ctx->socket;
+	struct socket *cl_socket;
 
-#if ( defined(_WIN32) || defined(_WIN32_WCE) ) && !defined(EFIX64) && \
-    !defined(EFI32)
-        if( WSAGetLastError() == WSAECONNRESET )
-            return( MBEDTLS_ERR_NET_CONN_RESET );
-#else
-        if( errno == EPIPE || errno == ECONNRESET )
-            return( MBEDTLS_ERR_NET_CONN_RESET );
+	ret = sock_create(PF_INET, SOCK_STREAM, IPPROTO_TCP, &cl_socket);
+	if(ret < 0)
+		return MBEDTLS_ERR_NET_ACCEPT_FAILED;
 
-        if( errno == EINTR )
-            return( MBEDTLS_ERR_SSL_WANT_READ );
+	ret = srv_socket->ops->accept(srv_socket, cl_socket, 0);
+	if (ret < 0) {
+		switch(ret) {
+#if defined EAGAIN
+		case -EAGAIN:
+#endif
+#if defined EWOULDBLOCK && EWOULDBLOCK != EAGAIN
+		case -EWOULDBLOCK:
 #endif
+			return MBEDTLS_ERR_SSL_WANT_READ;
+		}
+printk("%i\n", ret);
+		return MBEDTLS_ERR_NET_ACCEPT_FAILED;
+	}
 
-        return( MBEDTLS_ERR_NET_RECV_FAILED );
-    }
+	client_ctx->socket = cl_socket;
 
-    return( ret );
+	return 0;
 }
+EXPORT_SYMBOL(mbedtls_net_accept);
 
 /*
- * Read at most 'len' characters, blocking for at most 'timeout' ms
+ * Read at most 'len' characters
  */
-int mbedtls_net_recv_timeout( void *ctx, unsigned char *buf, size_t len,
-                      uint32_t timeout )
+int
+mbedtls_net_recv(void *ctx, unsigned char *buf, size_t len)
 {
-    int ret;
-    struct timeval tv;
-    fd_set read_fds;
-    int fd = ((mbedtls_net_context *) ctx)->fd;
-
-    if( fd < 0 )
-        return( MBEDTLS_ERR_NET_INVALID_CONTEXT );
-
-    FD_ZERO( &read_fds );
-    FD_SET( fd, &read_fds );
+	struct socket *sock = ((mbedtls_net_context *)ctx)->socket;
+	struct msghdr msg;
+	struct iovec iov;
+	mm_segment_t oldfs;
+	int ret;
+
+	if(sock == NULL)
+		return MBEDTLS_ERR_NET_INVALID_CONTEXT;
+
+	if(sock->sk == NULL)
+		return MBEDTLS_ERR_NET_INVALID_CONTEXT;
+
+	iov.iov_base = buf;
+	iov.iov_len = len;
+
+	memset(&msg, 0, sizeof(msg));
+	msg.msg_iter.type = ITER_IOVEC;
+	msg.msg_iter.count = len;
+	msg.msg_iter.iov = &iov;
+	msg.msg_iter.nr_segs = 1;
+
+	oldfs = get_fs();
+	set_fs(KERNEL_DS);
+	ret = sock_recvmsg(sock, &msg, len, MSG_DONTWAIT);
+	set_fs(oldfs);
+
+	if(ret < 0) {
+		switch(ret) {
+#if defined EAGAIN
+		case -EAGAIN:
+#endif
+#if defined EWOULDBLOCK && EWOULDBLOCK != EAGAIN
+		case -EWOULDBLOCK:
+#endif
```

Comment:
Don't you know what is defined? Usually such ifdefs are used in crossplathform code, not in our case.